### PR TITLE
Contract agent and deployer Interface

### PR DESCRIPTION
**Pull request by KPrasch**
- Separates the usage and deployment of NuCypher ethereum smart contracts, by introducing `Agent` and `Deployer` classes. This gives us a safer, more extensible, and layered way of managing the contracts
- Provides a python-client interface for Alice and Ursula to manage policies
- Reorganizes and updates tests; Extrapolates test utilities module
- Includes smart contract gas optimizations 

_Note: Includes changes from PR #8_

### CODE REVIEW

**Comment by KPrasch** [comment]
Amidst the KMS integration and ongoing contract updates, there are still many remaining TODOs in this PR, but I'd like to merge it in order to keep our working trees up to date without having to rebase these commits.

File: nkms_eth/blockchain.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
+from abc import ABC
+
 from nkms_eth.config import PopulusConfig
 
 
-class Blockchain:
+class TheBlockchain(ABC):
```
Comment (jMyles) [inline nkms_eth/blockchain.py]: Baha, awesome.

File: nkms_eth/config.py [hunk 2]
```diff
@@ -6,13 +7,81 @@
...
+        2 * 10 ** 7,
+        __max_awarded_periods,
+        __max_awarded_periods,
+        _min_release_periods,
+        __min_allowed_locked,
+        __max_allowed_locked
+    ]
+
+    class MinerInfoField(Enum):
+        MINERS_LENGTH = 0
```
Comment (jMyles) [inline nkms_eth/config.py]: Not sure I'm understanding here.  We apparently only use it once?  I hear you saying that we'll use it more in the future?

Comment (KPrasch) [reply nkms_eth/config.py]: Yes - it is possible to use in the future on the client layer, Nonetheless; contract tests use it more.  The intention is to provide a uniform/named way to refer to Miner structs on the contract instance.

File: nkms_eth/actors.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+class PolicyAuthor(TokenActor):
+    """Alice"""
+
+    def __init__(self, address: bytes, policy_agent):
+        super().__init__(token_agent=policy_agent._token, address=address)
+        self.policy_agent = policy_agent
+
+        self._arrangements = OrderedDict()    # Track authored policies by id
+
+    def make_arrangement(self, miner: Miner, periods: int, rate: int, arrangement_id: bytes=None) -> 'BlockchainArrangement':
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nkms_eth/actors.py]: I believe this is deprecated in a future commit, in relocating make arrangement to Policy

File: nkms_eth/agents.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+import random
+from abc import ABC
+from functools import partial
+from typing import Set, Generator, List
+
+from nkms_eth.deployers import MinerEscrowDeployer, NuCypherKMSTokenDeployer, PolicyManagerDeployer, ContractDeployer
+
+
+class EthereumContractAgent(ABC):
+    _principal_contract_name = NotImplemented
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nkms_eth/agents.py]: *hint hint*

File: nkms_eth/actors.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+
+class PolicyAuthor(TokenActor):
+    """Alice"""
+
+    def __init__(self, address: bytes, policy_agent):
+        super().__init__(token_agent=policy_agent._token, address=address)
+        self.policy_agent = policy_agent
+
+        self._arrangements = OrderedDict()    # Track authored policies by id
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply nkms_eth/actors.py]: Yep - better name.

**Pull request merged**